opm/autodiff/AquiferCarterTracy.hpp

```diff
@@ -0,0 +1,505 @@
+            std::vector<Scalar> coeff_;
+            
+
+            // We fit the tabular data using a polynomial fit
+            // Modified from Copyright (C) 2014  Clifford Wolf <clifford@clifford.at> 
+            // http://svn.clifford.at/handicraft/2014/polyfit/polyfit.cc
+            inline void polynomial_fit( const std::vector<Scalar> &X, const std::vector<Scalar> &y, 
+                                        std::vector<Scalar> &coeff, int order, bool bias) const
+            {
+                int colNum = (bias)? order + 1 : order;
+                Eigen::MatrixXd A(X.size(), colNum);
+                Eigen::VectorXd y_mapped = Eigen::VectorXd::Map(&y.front(), y.size());
+                Eigen::VectorXd result;
+
+                assert(X.size() == y.size());
+                assert(X.size() >= colNum);
+
+                // create matrix
+                for (size_t i = 0; i < X.size(); i++)
+                for (size_t j = 0; j < colNum; j++)
+                    A(i, j) = (bias)? pow(X.at(i), j) : pow(X.at(i), j+1);
+
+                // solve for linear least squares fit
+                result = A.householderQr().solve(y_mapped);
+
+                coeff.resize(colNum);
+                for (size_t i = 0; i < colNum; i++)
+                    coeff[i] = result[i];
+            }
+
+            inline void init_quantities(const Aquancon::AquanconOutput& connection)
+            {
+                // We reset the cumulative flux at the start of any simulation, so, W_flux = 0
+                W_flux_ = 0.;
+
+                // We next get our connections to the aquifer and initialize these quantities using the initialize_connections function
+                initialize_connections(connection);
+
+                // pa0_ is the initial aquifer water pressure. Must be calculated from equilibrium if left default,
+                // or we get the information from the deck Hacked to make it at 45e6 Pa
+                calculate_reservoir_equilibrium();
+                
+                pressure_previous_.resize(cell_idx_.size(), 0.);
+                pressure_current_.resize(cell_idx_.size(), 0.);
+
+                rhow_.resize(cell_idx_.size(), 998.0); 
+                Qai_.resize(cell_idx_.size(), 0.);
+
+                polynomial_fit(aqutab_td_, aqutab_pi_, coeff_, 2, true);
+            }
+
+            inline void get_current_Pressure_cell(std::vector<Scalar>& pressure_water, const int idx, const IntensiveQuantities& intQuants)
+            {
+                const auto& fs = intQuants.fluidState();
+                pressure_water[idx] = fs.pressure(FluidSystem::waterPhaseIdx).value();
+            }
+
+            inline void get_current_density_cell(std::vector<Scalar>& rho_water, const int idx, const IntensiveQuantities& intQuants)
+            {
+                const auto& fs = intQuants.fluidState();
+                rho_water[idx] = fs.density(FluidSystem::waterPhaseIdx).value();
```

Comment:
You can think if you want to keep the derivatives from the water density, it depends on the formulation you are actually using.